### DISCUSSION TITLE
Welcome to py4godot Discussions!

### DISCUSSION THREAD

**Discussion by niklas2902** [niklas2902/py4godot#21] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by TheBarrenDev** [comment]
I want to ask how I'm supposed to implement character movements using this code. I'm making an RPG, but I can't get the character to move. Here's the code I created after watching a YouTube tutorial:

```python
class Player(CharacterBody2D):

    # Define properties like this
    movement_speed: float = 100.0

    def _ready(self) -> None:
        pass
        # Put initialization code here

    def _process(self, delta: float) -> None:
        direction: Vector2 = Vector2.ZERO(0, 0)
        direction.x = Input.get_action_strength("right") - Input.get_action_strength("left")
        direction.y = Input.get_action_strength("down") - Input.get_action_strength("up")
        
        velocity = direction * movement_speed

    def _physics_process(self, delta: float) -> None:
        move_and_slide()
```

Forgive me if I've done something wrong—I'm still new to this!- also i need this to be done quick

**Reply by TheBarrenDev** [reply]
I think it didn't get the input, so the character doesn't move

**Reply by lazerstory** [reply]
Check if you've assigned your actions (right, left, down and up) to keys (WASD and/or arrows) by going to : Project > Project Settings > Input Map.
Your input map should look like this :
<img width="1187" height="359" alt="inputmap" src="https://github.com/user-attachments/assets/43711f44-0c46-48d2-903f-380db6415aad" />
Also, I'm pretty sure there's no need for "(0, 0)" after "Vector2.ZERO", as Vector2.ZERO is a constant in the class Vector2 and not a function. See [Vector2's documentation](https://docs.godotengine.org/en/stable/classes/class_vector2.html#constants)

There are several great tutorials included in Godot's documentation, for most things I'd recommend to look there before looking for Youtube tutorials. See [this link](https://docs.godotengine.org/en/stable/tutorials/2d/2d_movement.html) for a tutorial on a how to achieve character movement.

**Reply by TheBarrenDev** [reply]
thx!
I followed the tutorial from the link you gave, and I also checked my input map, but I still couldn't get the character moving. Here's the code:
class player(CharacterBody2D):
	movement_spd: float = 100.0

def _ready(self):
	pass
	
def _input(self, delta):
	input_direction = Input.get_vector("left", "right", "up", "down")
	velocity = input_direction * movement_spd

def _physics(self, delta):
	get_input()
	move_and_slide()

Again, sorry if I did something wrong

**Reply by niklas2902** [reply]
Hey, sorry about that. The API design isn’t ideal here. To access the `Input` singleton, you need to call the static `instance()` method on Input. It should look like this:

`Input.instance().get_vector("left", "right", "up", "down")`

For more details, check the developer guide: https://github.com/niklas2902/py4godot/wiki/Developer-Guide#singletons-autoload

**Reply by lazerstory** [reply]
Rather than "def _input(self, delta):", write "def get_input(self):". This is the function that you're calling in the "physics" function to update the player's velocity based on the input.

Also, since you're using Python and not GDScript, you should use "self" to access the class's attributes and functions. For example : "self.get_input()"
